### PR TITLE
[FreeSurface] Interface update

### PR DESCRIPTION
**Pull request by matekelemen**
**Description**
Update the interface of the FreeSurfaceApplication to conform to ```AnalysisStage``` and ```PythonSolver```. Simulations should be run using a ```MainKratos.py``` and ```ProjectParameters.json``` from now on.

**Changelog**
- add new ```EdgeBasedLevelSetSolver```
- remove old ```EdgeBasedLevelSetSolver```
- remove old ```free_surface_solver``` script
- add ```FreeSurfaceAnalysis```
- change tests to use the new interface

**Notes**
The solution step has a [conditional branch](https://github.com/KratosMultiphysics/Kratos/blob/9ef436761a66fefb49cb84ca0727b3a48af9426f/applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py#L267) which forces a recalculation if the step size is too large. However, this branch is never reached in the existing tests, so there's no way for me to properly test it.

During the solution loop, the solver has to perform a task ([Redistance](https://github.com/KratosMultiphysics/Kratos/blob/9ef436761a66fefb49cb84ca0727b3a48af9426f/applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py#L365)) with a specified frequency. However, the original script kept track multiple step counters that leads to a shift with respect to how ```PythonSolver``` counts steps. The current behaviour of the solver mirrors the original one because the test results depend greatly on when this task is carried out. That said, I think we should stick to the ```PythonSolver```'s step count and adjust the reference results accordingly.

Another possible minor issue with step counting is that the first 3 steps of the analysis aren't solved. I'm guessing the time integrator used here only needs 2 previous steps, but correct me if I'm wrong. The current behaviour reflects the original one (3 initial steps).

### CODE REVIEW

**Comment by philbucher** [comment]
uhh cool!
I have a bunch of comments, will post them in the next days

**Comment by matekelemen** [comment]
@philbucher uh-oh; It's the first time I took a look at the analysis pipeline so please go easy on me

File: applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py [hunk 1]
```diff
@@ -0,0 +1,439 @@
+# Kratos imports
```
Comment (rubenzorrilla) [inline applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: I'd keep the `levelset` keyword convention as we normally do in Kratos (this eases browsing the code). Note that the old solver used this.
Same for `edgebased`. All the the edge-based utilities have `edgebased` as keyword.

File: applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+        # Note:
+        # To stick to the original implementation, 3 initial steps are taken
+        # without solving the system, though I'd assume we only need 2.
+        while True and (3 < self.model_part.ProcessInfo[KratosMultiphysics.STEP]):
+            self.__SolveLocalSolutionStep()
+
+            # Check step size and revert if it's too large
+            # Note: does hardcoding a 0.95 safety factor make sense here
+            # or should it be variable?
+            step_size_estimate = self.__EstimateTimeStep(0.95, self.current_max_step_size)
```
Comment (ddiezrod) [inline applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: This 0.95 is the CFL, right? If so, I'd defintely have this as a solver setting. Just out of curiosity, does this need to be lower than 1 or you could use in principle a higher CFL (like 10) ?

Comment (antonialarese) [reply applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: Consider that this formulation is semi explicit this put more restrictions on CFL

File: applications/FreeSurfaceApplication/tests/generalTests.py [hunk 3]
```diff
@@ -10,192 +12,62 @@ def GetFilePath(fileName):
...
-                    fluid_solver.ReduceTimeStep(self.main_model_part, time)  # this is to set the database to the value at the beginning of the step
-
-                    safety_factor *= 3.00000e-01
-                    reduced_dt = fluid_solver.EstimateTimeStep(safety_factor, max_Dt)
+    def setUp(self):
+        self.check_tolerance = 1e-5
+        self.print_reference_values = False
 
-                    time = time - Dt + reduced_dt
+    def checkResults(self, reference_file_name: str):
```
Comment (philbucher) [inline applications/FreeSurfaceApplication/tests/generalTests.py:45]: I know this was already like this, but FYI there are better / less manual / standard solutions for this in the Core

File: applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+
+    def __Log(self, message: str) -> None:
+        KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, message)
+
+
+
+
+
+def CreateSolver(model: KratosMultiphysics.Model, parameters: KratosMultiphysics.Parameters) -> EdgeBasedLevelSetSolver:
+    if not issubclass(type(model), KratosMultiphysics.Model):
```
Comment (philbucher) [inline applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: usually we use `isinstance`

File: applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+        self.model_part.AddNodalSolutionStepVariable(KratosMultiphysics.VISCOSITY)
+        self.model_part.AddNodalSolutionStepVariable(KratosMultiphysics.DIAMETER)
+        self.model_part.AddNodalSolutionStepVariable(KratosMultiphysics.LIN_DARCY_COEF)
+        self.model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NONLIN_DARCY_COEF)
+        self.model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NODAL_AREA)
+        self.model_part.AddNodalSolutionStepVariable(KratosMultiphysics.STRUCTURE_VELOCITY)
+
+
+    def AddDofs(self) -> None:
+        for node in self.model_part.Nodes:
```
Comment (philbucher) [inline applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: usually we use the variable utils here

File: applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+import enum
+import pathlib
+
+
+class EdgeBasedLevelSetSolver(PythonSolver):
+
+    class PorousResistanceComputation(enum.IntEnum):
+        none   = 0
+        ergun  = 1
+        custom = 2
```
Comment (philbucher) [inline applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: maybe uppercase would be more applicable here

**Review by philbucher — COMMENTED** [review]
ok from my side!
Nice work!

**Comment by mpentek** [comment]
> This seems great! Thanks Máté. It seems perfectly aligned with what we spoke. I need to have a closer look at it but seems fine.

Dear @antonialarese please try to have a look so that we can have a (first) updated version merged.

**Review by philbucher — DISMISSED** [review]
Nice work :+1:

**Review by mpentek — DISMISSED** [review]
Thanks!

File: applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py (hunk 1 above)
Comment (rubenzorrilla) [reply applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: Please, take this into account before merging.

Comment (antonialarese) [reply applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: Please, @mpentek consider @rubenzorrilla comments on name conventions before merging

**Comment by antonialarese** [comment]
> > This seems great! Thanks Máté. It seems perfectly aligned with what we spoke. I need to have a closer look at it but seems fine.
> 
> @antonialarese we are going to proceed with a merge to be able to benefit from the updated version.

@mpentek THANKS! This is great.  Form my side you can proceed providing that you consider @rubenzorrilla  and @philbucher  comments

File: applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py (hunk 1 above)
Comment (matekelemen) [reply applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: @rubenzorrilla I'm working on updating the PR and will push changes with your and Philipp's suggestions + codacity fixes soon.

Comment (rubenzorrilla) [reply applications/FreeSurfaceApplication/python_scripts/edge_based_level_set_solver.py]: Perfect! Thanks for the update :wink:

**Comment by matekelemen** [comment]
Ready for approval; made the requested cosmetic changes.

File: applications/FreeSurfaceApplication/python_scripts/free_surface_analysis.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        # Base class constructor defines:
+        # - self.model
+        # - self.project_parameters
+        # - self.echo_level
+        # - self.parallel_type
+        AnalysisStage.__init__(self, model, parameters)
+
+
+    def _CreateSolver(self) -> EdgeBasedLevelSetSolver:
+        return EdgeBasedLevelSetSolver(self.model, self.project_parameters["solver_settings"].Clone())
```
Comment (philbucher) [inline applications/FreeSurfaceApplication/python_scripts/free_surface_analysis.py]:
```suggestion
        return EdgeBasedLevelSetSolver(self.model, self.project_parameters["solver_settings"])
```

this should not be done
the idea is to be able to see the full and validated params

File: applications/FreeSurfaceApplication/tests/Reservoir2D/ProjectParameters.json [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        "extrapolation_layers"          : 5,
+        "number_of_initial_time_steps"  : 10,
+        "initial_time_step_size"        : 1e-5,
+        "reduction_on_failure"          : 3e-1,
+        "stabdt_pressure_factor"        : 1e0,
+        "stabdt_convection_factor"      : 1e-2,
+        "tau2_factor"                   : 1e0,
+        "edge_detection_angle"          : 4.5e1,
+        "assume_constant_pressure"      : false,
+        "compute_porous_resistance_law" : "NONE",
```
Comment (philbucher) [inline applications/FreeSurfaceApplication/tests/Reservoir2D/ProjectParameters.json:30]: usually we use all lower case in the json, but I see where you are coming from

**Review by philbucher — DISMISSED** [review]
:+1:

**Comment by mpentek** [comment]
@matekelemen feel free to merge, any minor changes will come in subsequent PRs

**Review by philbucher — APPROVED** [review]

**Comment by mpentek** [comment]
@matekelemen please see if anything here is needed or it can be merged and closed.

**Pull request merged**